### PR TITLE
mullakhmetov / Chapter #2

### PR DESCRIPTION
**Pull request by mullakhmetov**

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +359,9 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList l r list
+  | l < 0 || r < 0 = []
```
Comment (chshersh) [inline src/Chapter2.hs:363]: You can cover extra case here when `l > r` to return an empty list earlier

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,7 +651,11 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [Int] -> [Int]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven [x, _] = [x]
```
Comment (chshersh) [inline src/Chapter2.hs:657]: This case is actually redundant, as it's covered by the next line.
Note, that `[x, _]` is syntax sugar for `(x : _ : []`).
Though, it is very helpful to cover smaller cases before generalizing :slightly_smiling_face:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -740,7 +775,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains x l = filter (elem x) l
```
Comment (chshersh) [inline src/Chapter2.hs:779]: You can actually eta-reduce this function as well

File: src/Chapter2.hs [hunk 4]
```diff
@@ -841,7 +879,13 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [Int] -> [Int]
+rotate _ [] = []
+rotate _ [x] = [x]
```
Comment (chshersh) [inline src/Chapter2.hs:884]: It looks like this case doesn't return an empty list for negative numbers :thinking:

File: src/Chapter2.hs [hunk 5]
```diff
@@ -841,7 +879,13 @@ list.
...
+rotate times l
+  | times < 0 = []
+  | times == 0 = l
+  | otherwise = subList times (times + length l - 1) (cycle l)
```
Comment (chshersh) [inline src/Chapter2.hs:888]: Well done on reusing the `subList` function in `rotate`! So far, you are the first one :1st_place_medal:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +901,10 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [Int] -> [Int]
+rewind [] = []
+rewind [x] = [x]
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:907]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
@mullakhmetov Nice work on solving tasks from Chapter Two!
And well done with the advanced task :+1: 

I left my feedback, but I'm going to approve this PR anyways, as solutions look good :100: 
Feel free to close this PR after changes, or just now if you don't want to spend too much time on this :slightly_smiling_face:

File: src/Chapter2.hs (hunk 6 above)
Comment (mullakhmetov) [reply src/Chapter2.hs]: Thanks for your feedback and encouraging me to recall SICP lessons about tail recursion :)

**Pull request closed without merging**